py-shared/tests/test_crc64.py

```diff
@@ -0,0 +1,135 @@
+"""Tests for CRC64-NVME checksum utilities."""
+
+import pytest
+
+from quilt_shared.crc64 import combine_crc64nvme, crc64_extend
+
+
+def test_extend_zero_bytes():
+    """Extending by 0 bytes returns unchanged CRC."""
+    crc = 0x123456789ABCDEF0
+    assert crc64_extend(crc, 0) == crc
+
+
+def test_extend_additive():
+    """Extending by (a + b) equals extending by a then b."""
+    crc = 0x123456789ABCDEF0
+    a, b = 100, 200
+
+    result_combined = crc64_extend(crc, a + b)
+    result_step = crc64_extend(crc64_extend(crc, a), b)
+
+    assert result_combined == result_step
+
+
+def test_extend_large_offset():
+    """Handles multi-GiB offsets efficiently (O(log n) performance)."""
+    crc = 0x1234567890ABCDEF
+    large_offset = 5 * 1024**3  # 5 GiB (AWS S3 max part size)
+    result = crc64_extend(crc, large_offset)
+    assert 0 <= result < (1 << 64)
+
+
+def test_extend_validation():
+    """Input validation: negative data_len, out-of-range CRC."""
+    # Negative data_len
+    with pytest.raises(ValueError, match="data_len must be non-negative"):
+        crc64_extend(0, -1)
+
+    # Invalid CRC (negative)
+    with pytest.raises(ValueError, match="crc must be 64-bit unsigned"):
+        crc64_extend(-1, 100)
+
+    # Invalid CRC (too large)
+    with pytest.raises(ValueError, match="crc must be 64-bit unsigned"):
+        crc64_extend(1 << 64, 100)
+
+
+def test_combine_empty_parts():
+    """Empty parts list returns zero CRC."""
+    assert combine_crc64nvme([], []) == b"\x00" * 8
+
+
+def test_combine_single_part():
+    """Single part returns unchanged."""
+    part_crc = b"\x12\x34\x56\x78\x9a\xbc\xde\xf0"
+    assert combine_crc64nvme([part_crc], [1024]) == part_crc
+
+
+def test_combine_two_parts():
+    """Combining two parts produces different result."""
+    part_crcs = [b"\x00\x00\x00\x00\x00\x00\x00\x01", b"\x00\x00\x00\x00\x00\x00\x00\x02"]
+    part_sizes = [8 * 1024 * 1024, 8 * 1024 * 1024]
+
+    result = combine_crc64nvme(part_crcs, part_sizes)
+
+    assert len(result) == 8
+    assert result != part_crcs[0]
+    assert result != part_crcs[1]
+
+
+def test_combine_order_matters():
+    """Part order affects the result."""
+    part_crcs = [b"\x00\x00\x00\x00\x00\x00\x00\x01", b"\x00\x00\x00\x00\x00\x00\x00\x02"]
+    part_sizes = [1024, 2048]
+
+    forward = combine_crc64nvme(part_crcs, part_sizes)
+    reversed = combine_crc64nvme([part_crcs[1], part_crcs[0]], [part_sizes[1], part_sizes[0]])
+
+    assert forward != reversed
+
+
+def test_combine_mismatched_lengths():
+    """Mismatched inputs raise ValueError."""
+    with pytest.raises(ValueError, match="must have the same length"):
+        combine_crc64nvme([b"\x00" * 8, b"\x01" * 8], [1024])
+
+
+def test_combine_many_parts():
+    """Handles 640 parts efficiently (~5 GiB file)."""
```

Comment:
how do we check it handles it efficiently?

Reply:
by ensuring it does not timeout lol. maybe we should add some timer idk to test executes under a second or smth. specific perf details are not that important, as long as it is "sensible"